### ISSUE TITLE
HPatches homography estimation

### ISSUE THREAD

**Issue by popovata** [rpautrat/SuperPoint#202]
Hello,
I tried to estimate a homography on HPatches dataset using pretrained model sp_v6. I got quite small correctnesses. Could you clarify, if I need to finetune a model before estimation?

**Comment by rpautrat** [comment]
Hi, you shouldn't need to fine-tune the sp_v6 model to do homography estimation on HPatches. What code/command did you use exactly? Note that you should pay extra care with image resizing (if you resize them), as the homography needs to be changed as well.

**Comment by popovata** [comment]
Thanks for the quick response! I use code from evaluations/descriptor_evaluation.py and from  match_features_demo.py.  I use function keep_shared_points(keypoint_map, H, keep_k_points=1000) from  evaluations/descriptor_evaluation.py and function for image preprocessing from match_features_demo.py. The function for tomography estimation is:


def homography_estimate(weights_name, img1_file, img2_file_path, real_h_path, keep_k_best=2000):
    img_size = (480, 640)
    real_H = np.loadtxt(real_h_path, usecols=range(3))
    weights_root_dir = Path("/home/tanya/bachelor/bachelor/SuperPoint/data_dir/saved_models")
    weights_root_dir.mkdir(parents=True, exist_ok=True)
    weights_dir = Path(weights_root_dir, weights_name)

    graph = tf.Graph()
    with tf.compat.v1.Session(graph=graph) as sess:
        tf.saved_model.loader.load(sess,
                                   [tf.saved_model.tag_constants.SERVING],
                                   str(weights_dir))

        input_img_tensor = graph.get_tensor_by_name('superpoint/image:0')
        output_prob_nms_tensor = graph.get_tensor_by_name('superpoint/prob_nms:0')
        output_desc_tensors = graph.get_tensor_by_name('superpoint/descriptors:0')

        img1, img1_orig = preprocess_image(img1_file, img_size)
        out1 = sess.run([output_prob_nms_tensor, output_desc_tensors],
                        feed_dict={input_img_tensor: np.expand_dims(img1, 0)})
        keypoint_map1 = np.squeeze(out1[0])
        descriptor_map1 = np.squeeze(out1[1])

        kp1 = keep_shared_points(keypoint_map1, real_H)
        desc1 = descriptor_map1[kp1[:, 0], kp1[:, 1]]

        img2, img2_orig = preprocess_image(img2_file_path, img_size)
        out2 = sess.run([output_prob_nms_tensor, output_desc_tensors],
                        feed_dict={input_img_tensor: np.expand_dims(img2, 0)})
        keypoint_map2 = np.squeeze(out2[0])
        descriptor_map2 = np.squeeze(out2[1])
        kp2 = keep_shared_points(keypoint_map2, np.linalg.inv(real_H))
        desc2 = descriptor_map2[kp2[:, 0], kp2[:, 1]]
        matches = match_descriptors(desc1, desc2)  #I use cv2.BFMatcher here

        matches_idx = np.array([m.queryIdx for m in matches])
        m_keypoints = kp1[matches_idx, :]
        matches_idx = np.array([m.trainIdx for m in matches])
        m_warped_keypoints = kp2[matches_idx, :]

        H, inliers = compute_homography(m_keypoints, m_warped_keypoints)
        corners = np.array([[0, 0, 1],
                                [keypoint_map1.shape[1] - 1, 0, 1],
                                [0, keypoint_map1.shape[0] - 1, 1],
                                [keypoint_map1.shape[1] - 1, keypoint_map1.shape[0] - 1, 1]])
        real_warped_corners = np.dot(corners, np.transpose(real_H))
        real_warped_corners = real_warped_corners[:, :2] / real_warped_corners[:, 2:]
        warped_corners = np.dot(corners, np.transpose(H))
        warped_corners = warped_corners[:, :2] / warped_corners[:, 2:]
        mean_dist = np.mean(np.linalg.norm(real_warped_corners - warped_corners, axis=1))
        correctness = float(mean_dist <= 5)
        return correctness

**Comment by rpautrat** [comment]
Yes, so it seems that the issue comes from your image resizing, as I expected. If I am not wrong, you resize the images to 480x640, but keep the original homography. The homography should however be adapted to the new image size, for example with the following function: https://github.com/rpautrat/SuperPoint/blob/313a0a115961bc2a98a5ed1d7bde33fa99542398/superpoint/datasets/patches_dataset.py#L62

**Comment by popovata** [comment]
Thank you for your time. Yes, I really didn't change the homography. I've fixed it, but the correctness doesn't seem better.

I adapt the homography with the function you notice above, here is my code again:
def homography_estimate(weights_name, img1_file, img2_file_path, real_h_path, keep_k_best=2000):
    img_size = (480, 640)
    real_H = np.loadtxt(real_h_path, usecols=range(3))

    #homographies = tf.data.Dataset.from_tensor_slices(real_H)

    weights_root_dir = Path("/home/tanya/SuperPoint/exper/saved_models")
    weights_root_dir.mkdir(parents=True, exist_ok=True)
    weights_dir = Path(weights_root_dir, weights_name)

    graph = tf.Graph()
    with tf.compat.v1.Session(graph=graph) as sess:
        tf.saved_model.loader.load(sess,
                                   [tf.saved_model.tag_constants.SERVING],
                                   str(weights_dir))

        input_img_tensor = graph.get_tensor_by_name('superpoint/image:0')
        output_prob_nms_tensor = graph.get_tensor_by_name('superpoint/prob_nms:0')
        output_desc_tensors = graph.get_tensor_by_name('superpoint/descriptors:0')

        img1, img1_orig = preprocess_image(img1_file, img_size)
        out1 = sess.run([output_prob_nms_tensor, output_desc_tensors],
                        feed_dict={input_img_tensor: np.expand_dims(img1, 0)})
        keypoint_map1 = np.squeeze(out1[0])
        descriptor_map1 = np.squeeze(out1[1])

        img2, img2_orig = preprocess_image(img2_file_path, img_size)
        out2 = sess.run([output_prob_nms_tensor, output_desc_tensors],
                        feed_dict={input_img_tensor: np.expand_dims(img2, 0)})
        keypoint_map2 = np.squeeze(out2[0])
        descriptor_map2 = np.squeeze(out2[1])

        real_H = tf.Session().run(_adapt_homography_to_preprocessing(real_H, img1_orig.shape[:2], img2_orig.shape[:2], img_size))

        kp1 = keep_shared_points(keypoint_map1, real_H)
        desc1 = descriptor_map1[kp1[:, 0], kp1[:, 1]]

        kp2 = keep_shared_points(keypoint_map2, np.linalg.inv(real_H))
        desc2 = descriptor_map2[kp2[:, 0], kp2[:, 1]]
        matches = match_descriptors(desc1, desc2)  #I use cv2.BFMatcher here

        matches_idx = np.array([m.queryIdx for m in matches])
        m_keypoints = kp1[matches_idx, :]
        matches_idx = np.array([m.trainIdx for m in matches])
        m_warped_keypoints = kp2[matches_idx, :]

        H, inliers = compute_homography(m_keypoints, m_warped_keypoints)
        corners = np.array([[0, 0, 1],
                                [keypoint_map1.shape[1] - 1, 0, 1],
                                [0, keypoint_map1.shape[0] - 1, 1],
                                [keypoint_map1.shape[1] - 1, keypoint_map1.shape[0] - 1, 1]])
        real_warped_corners = np.dot(corners, np.transpose(real_H))
        real_warped_corners = real_warped_corners[:, :2] / real_warped_corners[:, 2:]
        warped_corners = np.dot(corners, np.transpose(H))
        warped_corners = warped_corners[:, :2] / warped_corners[:, 2:]
        mean_dist = np.mean(np.linalg.norm(real_warped_corners - warped_corners, axis=1))
        print(mean_dist)
        correctness = float(mean_dist <= 5)
        return correctness

**Comment by rpautrat** [comment]
Did you try visualizing a couple images with the detected keypoints on it, to check that everything was fine? You can also try to warp the resized image with the adapted homography and check that you get the correct warped image.

Another potential source of mistake is in your image preprocessing. Do you normalize the images in [0, 1]?

**Comment by popovata** [comment]
Yes, I normalize the images by using   img_preprocessed = image / 255.

Visualization of keypoint detection and matching seems fine. 
![matches](https://user-images.githubusercontent.com/50377523/114632303-dc213e80-9cd7-11eb-9d99-5810010c3869.png)


I tried to warp the resized image with the adapted homography, I got this warped image
![warp](https://user-images.githubusercontent.com/50377523/114632804-daa44600-9cd8-11eb-9e72-ef45c23d60d5.png)

I just use the code you noticed above.
`def _adapt_homography_to_preprocessing(hom, shape, w_shape, size):
    H = tf.cast(hom, tf.float32)
    source_size = tf.cast(shape, tf.float32)
    source_warped_size = tf.cast(w_shape, tf.float32)
    target_size = tf.cast(tf.convert_to_tensor(size),
                          tf.float32)

    # Compute the scaling ratio due to the resizing for both images
    s = tf.reduce_max(tf.divide(target_size, source_size))
    up_scale = tf.diag(tf.stack([1. / s, 1. / s, tf.constant(1.)]))
    warped_s = tf.reduce_max(tf.divide(target_size, source_warped_size))
    down_scale = tf.diag(tf.stack([warped_s, warped_s, tf.constant(1.)]))

    # Compute the translation due to the crop for both images
    pad_y = tf.to_int32(((source_size[0] * s - target_size[0]) / tf.constant(2.0)))
    pad_x = tf.to_int32(((source_size[1] * s - target_size[1]) / tf.constant(2.0)))
    translation = tf.stack([tf.constant(1), tf.constant(0), pad_x,
                            tf.constant(0), tf.constant(1), pad_y,
                            tf.constant(0), tf.constant(0), tf.constant(1)])
    translation = tf.to_float(tf.reshape(translation, [3, 3]))
    pad_y = tf.to_int32(((source_warped_size[0] * warped_s - target_size[0])
                         / tf.constant(2.0)))
    pad_x = tf.to_int32(((source_warped_size[1] * warped_s - target_size[1])
                         / tf.constant(2.0)))
    warped_translation = tf.stack([tf.constant(1), tf.constant(0), -pad_x,
                                   tf.constant(0), tf.constant(1), -pad_y,
                                   tf.constant(0), tf.constant(0), tf.constant(1)])
    warped_translation = tf.to_float(tf.reshape(warped_translation, [3, 3]))

    H = warped_translation @ down_scale @ H @ up_scale @ translation
    return H`

**Comment by rpautrat** [comment]
The last image indicates that the homography is probably wrong. This might be due to the fact the the `adapt_homography_to_preprocessing` that I gave you is only adapted for a specific kind of image resizing, the one we are using in this repo: https://github.com/rpautrat/SuperPoint/blob/313a0a115961bc2a98a5ed1d7bde33fa99542398/superpoint/datasets/utils/pipeline.py#L94
If you did a different resizing in your `preprocess_image`, you need to adapth the homography differently. Or you can directly use the same image resizing as in this repo.

**Comment by popovata** [comment]
I tried to run export_descriptors.py  with sp_v6 model to understand how it should work.
I run:  python export_descriptors.py configs/superpoint_hpatches.config sp_v6 --export_name=sp_v6-hpatches-v
but got this error: 
![Screenshot at 2021-04-14 19:11:34](https://user-images.githubusercontent.com/50377523/114751380-88adff80-9d6e-11eb-9213-269d491e0633.png)
When I try:  python export_descriptors.py configs/superpoint_hpatches.config saved_models/sp_v6 --export_name=sp_v6-hpatches-v  I get this one:
![Screenshot at 2021-04-14 19:13:49](https://user-images.githubusercontent.com/50377523/114751759-0114c080-9d6f-11eb-9946-89f5e345b179.png)
What am I doing wrong?

**Comment by rpautrat** [comment]
Hi, which sp_v6 are you using, the exported model in the `pretrained_models` folder, or the [one provided in the Readme](https://drive.google.com/file/d/18VpMXubyWL6WoNfyq6kQ6g4AM19QJ0FV/view?usp=sharing)? You should use the latter with `export_descriptors.py`

**Comment by popovata** [comment]
Yes, I used the first one, thank you! I fixed it, now I got this error:
![Screenshot at 2021-04-14 22:19:28](https://user-images.githubusercontent.com/50377523/114773766-e3545500-9d88-11eb-9c14-23c69e3e483b.png)

**Comment by rpautrat** [comment]
Do you have an Nvidia GPU compatible with TensorFlow? It seems that Tensorflow cannot locate it on your computer...

**Comment by popovata** [comment]
Is it possible to run the program without Nvidia GPU?

**Comment by rpautrat** [comment]
Currently not, unfortunately... But it would also be pretty slow on CPU.

**Comment by popovata** [comment]
Hi! I installed tensorflow-gpu by steps in #173 .   My tf version is 1.14.0, but still Tensorflow cannot locate any GPU on my computer.  
When I try nvidia-smi, then I get this error: NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.

But conda installed all related packages.

Could you help me with solving this problem?

Setup:
Ubuntu 16.04
GeForce 840M

**Comment by rpautrat** [comment]
Did you install any NVIDIA driver then? Conda doesn't handle that. Here is the page to download the Nvidia driver: https://www.nvidia.com/download/index.aspx?lang=en-us
Alternatively, you can use the Ubuntu package manager to install the drivers, with ```sudo apt-get install nvidia-driver-XXX```, where XXX refers to the driver version.

**Comment by popovata** [comment]
I don't know why didn't have any NVIDIA driver installed.
I installed it, now it works great.
Thank you a lot:)
